client/src/components/profile/components/HeatMap.js

```diff
@@ -87,27 +128,124 @@ function HeatMap({ calendar, streak }) {
               'data-tip': `<b>${valueCount}</b> ${dateFormatted}`
             };
           }}
-          values={calendarData}
+          values={dataToDisplay}
         />
         <ReactTooltip className='react-tooltip' effect='solid' html={true} />
+
+        <Spacer />
+        <Row>
+          <div className='streak-container'>
+            <span className='streak'>
+              <b>Longest Streak:</b> {longestStreak || 0}
+            </span>
+            <span className='streak'>
+              <b>Current Streak:</b> {currentStreak || 0}
+            </span>
+          </div>
+        </Row>
+        <hr />
       </FullWidthRow>
-      <Spacer />
-      <FullWidthRow>
-        <div className='streak-container'>
-          <span className='streak'>
-            <b>Longest Streak:</b> {streak.longest || 0}
-          </span>
-          <span className='streak'>
-            <b>Current Streak:</b> {streak.current || 0}
-          </span>
-        </div>
-      </FullWidthRow>
-      <Spacer />
-      <hr />
-    </FullWidthRow>
-  );
+    );
+  }
 }
 
+HeatMapInner.propTypes = innerPropTypes;
+
+const HeatMap = props => {
+  const { calendar } = props;
+
+  /**
+   *  the following logic creates the data for the heatmap
+   *  from the users calendar and calculates their streaks
+   */
+
+  // create array of timestamps and turn into milliseconds
+  const timestamps = Object.keys(calendar).map(stamp => stamp * 1000);
+  const startOfTimestamps = startOfDay(new Date(timestamps[0]));
+  let endOfCalendar = startOfDay(Date.now());
+  let startOfCalendar;
+
+  // creates pages for heatmap
+  let pages = [];
+
+  do {
+    startOfCalendar = addDays(addMonths(endOfCalendar, -6), 1);
+
+    const newPage = {
+      startOfCalendar: startOfCalendar,
+      endOfCalendar: endOfCalendar
+    };
+
+    pages.unshift(newPage);
+
+    endOfCalendar = addDays(startOfCalendar, -1);
+  } while (startOfTimestamps < startOfCalendar);
+
+  let calendarData = [];
+  let dayCounter = pages[0].startOfCalendar;
+
+  // create an object for each day of the calendar period
+  while (dayCounter <= pages[pages.length - 1].endOfCalendar) {
+    // this is the format needed for react-calendar-heatmap
+    const newDay = {
+      date: startOfDay(dayCounter),
+      count: 0
+    };
+
+    calendarData.push(newDay);
+    dayCounter = addDays(dayCounter, 1);
+  }
+
+  let longestStreak = 0;
+  let currentStreak = 0;
+  let lastIndex = -1;
+
+  // add a point to each day with a completed timestamp and calculate streaks
+  timestamps.forEach(stamp => {
+    const index = calendarData.findIndex(day =>
+      isEqual(day.date, startOfDay(stamp))
+    );
+
+    if (index >= 0) {
+      // add one point for today
+      calendarData[index].count++;
+
+      // if timestamp is on a new day, deal with streaks
+      if (index !== lastIndex) {
```

Comment:
This looks good to me, but I think it would be worth creating a test or two. Just to verify that we get the expected current and longest streaks.

Reply:
I can try and add a few more at some point today.